### PR TITLE
upipe-freetype: add colors support

### PR DESCRIPTION
**Pull request by quarium**

### CODE REVIEW

File: lib/upipe-freetype/upipe_freetype.c [hunk 1]
```diff
@@ -489,6 +580,117 @@ static void upipe_freetype_input(struct upipe *upipe, struct uref *uref,
...
+}
+
+/** @internal @This converts rgba color to yuva.
+ *
+ * @param upipe description structure of the pipe
+ * @param rgba rgb and alpha color to convert
+ * @param yuva yuv and alpha color converted
+ * @return an error code
+ */
+static int upipe_freetype_rgba_to_yuva(struct upipe *upipe,
```
Comment (cmassiot) [inline lib/upipe-freetype/upipe_freetype.c]: If this not specific to freetype, it should maybe be in ubuf_pic.c.

File: lib/upipe-freetype/upipe_freetype.c [hunk 2]
```diff
@@ -109,6 +109,14 @@ struct upipe_freetype {
     /** baseline right offset */
     int64_t yoff;
 
+    /** full range */
+    bool fullrange;
+
+    /** background color YUV */
```
Comment (funman) [inline lib/upipe-freetype/upipe_freetype.c]: YUVA ?

File: lib/upipe-freetype/upipe_freetype.c [hunk 3]
```diff
@@ -254,6 +262,19 @@ static struct upipe *upipe_freetype_alloc(struct upipe_mgr *mgr,
     upipe_freetype->yoff = 0;
     upipe_freetype->font = NULL;
     upipe_freetype->pixel_size = 0;
+    upipe_freetype->fullrange = false;
+    if (ubase_check(uref_pic_flow_get_full_range(flow_def)))
+        upipe_freetype->fullrange = true;
```
Comment (funman) [inline lib/upipe-freetype/upipe_freetype.c]: upipe_freetype->fullrange = ubase_check(uref_pic_flow_get_full_range(flow_def)); ?

File: lib/upipe-freetype/upipe_freetype.c (hunk 1 above)
Comment (quarium) [reply lib/upipe-freetype/upipe_freetype.c]: Fixed

File: lib/upipe-freetype/upipe_freetype.c (hunk 2 above)
Comment (quarium) [reply lib/upipe-freetype/upipe_freetype.c]: Fixed

File: lib/upipe-freetype/upipe_freetype.c (hunk 3 above)
Comment (quarium) [reply lib/upipe-freetype/upipe_freetype.c]: Fixed

File: lib/upipe-freetype/upipe_freetype.c [hunk 4]
```diff
@@ -489,6 +579,94 @@ static void upipe_freetype_input(struct upipe *upipe, struct uref *uref,
...
 }
 
+/** @internal @This parses an RGBA color.
+ *
+ * @param upipe description structure of the pipe
+ * @param value RGBA value to parse
+ * @param rgba array filled with the r, g, b and a value
+ * @return an error code
+ */
+static int upipe_freetype_parse_rgba(struct upipe *upipe,
```
Comment (cmassiot) [inline lib/upipe-freetype/upipe_freetype.c]: Perhaps this too should be in ubuf_pic ?

Comment (quarium) [reply lib/upipe-freetype/upipe_freetype.c]: Fixed

**Pull request merged**